src/PrestaShopBundle/EventListener/AccessDeniedListener.php

```diff
@@ -0,0 +1,116 @@
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+namespace PrestaShopBundle\EventListener;
+
+use Symfony\Component\HttpKernel\Event\GetResponseForExceptionEvent;
+use Symfony\Component\Security\Core\Exception\AccessDeniedException;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use PrestaShopBundle\Security\Annotation\BetterSecurity;
+use Symfony\Component\Translation\TranslatorInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * Allow a redirection to the right url when using BetterSecurity annotation.
+ */
+class AccessDeniedListener
+{
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var SessionInterface
+     */
+    private $session;
+
+    public function __construct(RouterInterface $router, TranslatorInterface $translator, SessionInterface $session)
+    {
+        $this->router = $router;
+        $this->translator = $translator;
+        $this->session = $session;
+    }
+
+    /**
+     * @param GetResponseForExceptionEvent $event
+     */
+    public function onKernelException(GetResponseForExceptionEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        if (!$event->getException() instanceof AccessDeniedException) {
+            return;
+        }
+
+        $security = $event->getRequest()->get('_security');
+
+        if (empty($security) || !$security[0] instanceof BetterSecurity) {
+            return;
+        }
+
+        $event->allowCustomResponseCode();
+
+        $betterSecurity = $security[0];
+        $this->throwNotificationMessage($betterSecurity);
+        $url = $this->computeUrl($betterSecurity);
+
+        $event->setResponse(new RedirectResponse($url));
+    }
+
+    /**
+     * Compute the url for the redirection.
+     * @param BetterSecurity $betterSecurity
+     *
+     * @return string
+     */
+    private function computeUrl(BetterSecurity $betterSecurity)
+    {
+        $route = $betterSecurity->getRoute();
+        if ($route !== null) {
+            return $this->router->generate($route);
+        }
+
+        return $betterSecurity->getUrl();
+    }
+
+    /**
+     * Send an error message when redirected, will only work on migrated pages.
+     * @param BetterSecurity $betterSecurity
+     *
+     */
+    private function throwNotificationMessage(BetterSecurity $betterSecurity)
+    {
+        $this->session->getFlashBag()->add('error', $this->translator->trans($betterSecurity->getMessage(), [], $betterSecurity->getDomain()));
```

Comment:
Line too long:
```php
$this->session->getFlashBag()->add(
    'error',
    $this->translator->trans(
        $betterSecurity->getMessage(),
        [],
        $betterSecurity->getDomain()
    )
);
```

Reply:
coding style etc... I'll launch php-cs-fixer on this file.